### ISSUE TITLE
Randomized Search

### ISSUE THREAD

**Issue by pGit1** [aigamedev/scikit-neuralnetwork#123]
Is anyone else having trouble with Randomized Search with this package? I can see that the process kicks off but it takes an inordinate amount of time to process. In fact Ive had to stop the process many times becuase it seems to never stop. Even on 100 record datasets with cv=2...

Is it just me. I could not even do randomization with relatively simple networks. I found a work arond but it's still annoying. These other pul requests are awesome by the way. This could end up being the best package once it is all said and done with.

**Comment by alexjc** [comment]
Thanks, I've scheduled some time soon to add, test and document those various little features for the next revision.

For randomized search, are you doing multiple threads in parallel somehow (not familiar with sklearn here)? That may cause issues with Theano. If you pick a model that does, say, only 1 iteration quickly does it still get stuck?  Can you change the number of parallel threads to 1?

**Comment by pGit1** [comment]
Can't wait Alex! Concerning the parallel threads, how do I tweak this parameter?

**Comment by alexjc** [comment]
I don't know about threads, it's within sklearn that you may be able to set this. Just know that the underlying Theano code may not be able to deal with threads very well (there's a global compile lock) and it won't do well with multiple threads competing for GPU either.

**Comment by pGit1** [comment]
Alex here is my code:

from sklearn.datasets import make_classification
from sknn.mlp import  Layer, Regressor, Classifier as skClassifier 
import numpy as np
import scipy as sp 
import pandas as pd
from sklearn.grid_search import RandomizedSearchCV

# Load data set and target values
data, target = make_classification(n_samples=2500,
                                   n_features=45,
                                   n_informative=12,
                                   n_redundant=7)

max_net = skClassifier(layers= [Layer("Rectifier",units=10),
                                       Layer("Rectifier",units=10),
                                       Layer("Rectifier",units=10),
                                       Layer("Softmax")])
params={'learning_rate': sp.stats.uniform(0.0001, 0.05),
        'hidden0__units': sp.stats.randint(4, 20),
        'hidden0__type': ["Rectifier"],
        'hidden1__units': sp.stats.randint(4, 20),
        'hidden1__type': ["Rectifier"],
        'hidden2__units': sp.stats.randint(4, 20),
        'hidden2__type': ["Rectifier"],
        'batch_size':sp.stats.randint(10,1000),
        'learning_rule':["adagrad"]}
max_net2 = RandomizedSearchCV(max_net,param_distributions=params,n_iter=1,cv=2,scoring='f1',verbose=100,n_jobs=1,\
                             pre_dispatch=None)
max_net2.fit(data,target)

**Comment by alexjc** [comment]
`n_jobs` is set to 1 so probably using one thread only. I'll have to look into it.

**Comment by pGit1** [comment]
Thanks Alex. I hope you find something!

**Comment by alexjc** [comment]
If you interrupt it with CTRL+C, what callstack does it give you?

**Comment by pGit1** [comment]
Not sure but how to get to call stack. Control C didn't do anything when I
pressed it.

**Comment by alexjc** [comment]
So it just hangs completely? How do you exit?

**Comment by pGit1** [comment]
I just hit the stop button on the ipy notebook. How do I get a call stack?
Now that you have the code can you try?

**Comment by alexjc** [comment]
Ah, if you do it from the console outside of IPython it would give you a callstack (in a real file) when pressing CTRL+C. I will try it eventually but not immediately, more information is better until then!

**Comment by pGit1** [comment]
I did control+C in a regular pyhton module and it didn't return a call
stack. It just returned links to all the code that was being processed
during training.

**Comment by alexjc** [comment]
You need to set n_iter to something smaller or it will train a while until it reaches stability. If you train on CPU that can take a while. With `n_iter=5` the code exists quickly with no problems.

Closing this issue, let me know!
